### DISCUSSION TITLE
How to debug when a test failed

### DISCUSSION THREAD

**Discussion by LZhaoM** [vim/vim#16290] [Q&A] [answered]
Hi,

I'm trying to fix the vim tests on GNU Hurd. One of the tests failed is `Test_conceal_resize_02.dump`. I pasted the expected output and actual output below:

Expected:
```
  |o+0&#ffffff0|n|e| |t|w|o| |t|h|r|e@1| |f|o|u|r| |f|i|v|e|,| |t|h|e| >b|a|c|k|t|i|c|k|s| |s|h|o|u|l|d| |b|e| |c|o|n|c|e|a|l|e|d| @    16
  |~+0#4040ff13&| @73
  |~| @73
  |~| @73
  |~| @73
  |~| @73
  | +0#0000000&@56|1|,|4|0| @9|A|l@1|
```

Actual:
```
  |o+0&#ffffff0|n|e| |t|w|o| |t|h|r|e@1| |f|o|u|r| |f|i|v|e|,| |t|h|e| >b|a|c|k|t|i|c|k|s| |s|h|o|u|l|d| |b|e| |c|o|n|c|e|a|l|e|d| @    16
  |~+0#4040ff13&| @73
  |~| @73
  |~| @73
  |~| @73
  | +0#0000000&@56|1|,|4|0| @9|A|l@1|
  @75
```

I'm quite curious about how to track down where's the issue (it's not the issue of vim, the test failed only on Hurd). I haven't found a documentation talking about this (maybe I just missed it).

Any help is highly appreciated!

**Comment by chrisbra** [comment]
There is an additional `~| @73` line in the expected output. Looks like for some reason the screen window is smaller? 
BTW: you can use the `term_dumpdiff()` function to view the difference more easily. The test output should show the call to `term_dumpdiff()` call from the `testdir` directory.

**Reply by LZhaoM** [reply]
Hi @chrisbra,

Thanks for your reply.

I'm running Hurd in qemu on my laptop. I have an external monitor. I just unplugged the monitor and run `make test_conceal`. It gave me the following error without running the test:
```
From messages:
Screen size too small! Tests require at least 24 lines with 80 characters, got 22 lines with 104 characters
```

I then plugged my external monitor and run `make test_conceal` again, the tests started running. I think the following is what you called `term_dumpdiff()`:
```
1 FAILED:
Found errors in Test_conceal_resize_term():
Run 1, 16:54:24 - 16:54:29 in  5.150000 seconds:
command line..script /home/1speaker/apt-source/vim/vim-9.1.0861/src/testdir/runtest.vim[613]..function RunTheTest[57]..Test_conceal_resize_term[14]..VerifyScreenDump line 80: See dump file difference: call term_dumpdiff("testdir/failed/Test_conceal_resize_02.dump", "testdir/dumps/Test_conceal_resize_02.dump"); difference in line 6: "| +0#0000000&@56|1|,|4|0| @9|A|l@1| "; difference in line 7: "@75"
```

To be honest it's quite confused for me. Does it make sense to disable this test? If yes, How? Or does it indicate any issue on Hurd?

**Reply by chrisbra** [reply]
yes, from the src-directory, please run `vim --clean -c 'call term_dumpdiff("testdir/failed/Test_conceal_resize_02.dump", "testdir/dumps/Test_conceal_resize_02.dump")`. That should load the failing screen dump test and allows you to inspect the diff. See also the explanation at `:h term_dumpdiff()`.

**Reply by LZhaoM** [reply]
Wow this function is really useful, here is the output:
```
one two three four five, the backticks should be concealed                 
~                                                                          
~                                                                          
~                                                                          
~                                                                          
                                                         1,40          All 
                                                                           
=============== testdir/failed/Test_conceal_resize_02.dump ===============
                                                                           
                                                                           
                                                                           
                                                                           
                                                                           
XffffffffffffffffffffffffffffffffffffffffffffffffffffffffXXXXffffffffffXXXf
                                                         XXXX          XXX 
================ testdir/dumps/Test_conceal_resize_02.dump ================
one two three four five, the backticks should be concealed                 
~                                                                          
~                                                                          
~                                                                          
~                                                                          
~                                                                          
                                                         1,40          All 

```

**Reply by chrisbra** [reply]
That confirms, that in your output, there is one line missing of those filler lines after the actual buffer lines, those showing the `~`. It seems like the resizing does not work correctly. Can you run the test manually and especially check what happens on the `wincmd +`? See here:
https://github.com/vim/vim/blob/e62d93ead10b4c5818e3c0b7551f1784d24bfe33/src/testdir/test_conceal.vim#L239-L244

**Reply by LZhaoM** [reply]
Not sure how to run `  let buf = RunVimInTerminal('-S XTest_conceal_resize', {'rows': 6})` manually, but I didn't find anything wrong with `wincmd+` ( i.e. `wincmd+` increments `~` when there is space).

**Reply by chrisbra** [reply]
If you start vim from the `testdir` directory, e.g. `vim test_conceal.vim`, you can simply `:source` this file and all included dependencies. Then `RunVimInTerminal()` is automatically defined and you can simply call it interactively.

**Reply by LZhaoM** [reply]
Cool. I manually run the `call win_execute(buf->win_findbuf()[0], 'wincmd +') ` on two OS (Hurd and GNU/Linux). The issue I can see on Hurd is that the number of `~` does not increases when the command is executed. Probably I should go more in depth into the source code?

EDIT: more info: the size of window increases, but the number of `~` does not increase.

**Reply by chrisbra** [reply]
Hm, that is very strange. If the `~` lines doe not increase, what happens with the increase, e.g. the additional line that has been added? Probably need to run it via a debugger, to see what is going wrong then.

**Reply by LZhaoM** [reply]
Do you mean the debugging mentioned in the following? I found another debug method mentioned in `:h debug`, but I think that one is for debugging vim script.

https://github.com/vim/vim/blob/2e252474c4df5018b9819d86ebb70bf3b1b1a1af/src/INSTALL#L74-L80

**Reply by chrisbra** [reply, marked as answer]
Yes, `:h debug` is for debugging Vim Script. I would compile with `-g -DDEBUG` and then attach `gdb` to the inner Vim process, breaking at `win_setheight()` and see if you can find out what is going wrong.

**Reply by LZhaoM** [reply]
Thanks, mark this as the answer, if there is any further issues I will open another discussion :).

**Reply by puremourning** [reply]
A very long time ago, I put together [a tutorial on using vimspector to debug vim tests](https://puremourning.github.io/vimspector-web/demo-setup.html)

**Reply by LZhaoM** [reply]
@puremourning That's cool. Good to know. Thanks
